### PR TITLE
SIL: add the self-parameter to the list of type-dependent operands if…

### PR DESCRIPTION
**Pull request by eeckstein**
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

… an instruction uses the dynamic-self type.

This establishes a real def-use relation from the self-parameter to any instruction which uses the dynamic-self type.
This is an addition to what was already done for opened archetypes.
The biggest part of this commit is to rename "OpenedArchetypeOperands" to "TypeDependentOperands" as this name is now more appropriate.

Other than that the change includes:
*) type-dependent operands are now printed after a SIL instruction in a comment as "type-defs:" (for debugging)
*) FuncationSignatureOpts doesn't need to explicitly check if a function doesn't bind dynamic self to remove a dead self metadata argument
*) the check if a function binds dynamic self (used in the inliner) is much simpler now
*) also collect type-dependent operands for ApplyInstBase::SubstCalleeType and not only in the substitution list
*) with this SILInstruction::mayHaveOpenedArchetypeOperands (used in CSE) is not needed anymore and removed
*) fixed a bug in the cloner: SILBuilderWithPostProcess now inherits the OpenedArchetypesTracker from it's original builder
*) add type dependent operands to dynamic_method instruction

Regarding the generated code it should be a NFC.

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci Please test

**Comment by eeckstein** [comment]
@swift-ci Please test

**Comment by eeckstein** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
Excellent!

**Comment by eeckstein** [comment]
@swift-ci Please smoke test OS X platform

File: include/swift/SIL/SILCloner.h [hunk 1]
```diff
@@ -270,6 +270,7 @@ class SILBuilderWithPostProcess : public SILBuilder {
     {
       setInsertionPoint(SC.getBuilder().getInsertionBB(),
                         SC.getBuilder().getInsertionPoint());
+      setOpenedArchetypesTracker(SC.getBuilder().getOpenedArchetypesTracker());
```
Comment (swiftix) [inline include/swift/SIL/SILCloner.h]: @eeckstein Can you submit this fix a separate commit? I think we may also take it into the swift-3.0-branch.

**Comment by eeckstein** [comment]
I created another PR, because the smoke test doesn't start for this one.
https://github.com/apple/swift/pull/4277

**Comment by najacque** [comment]
@eeckstein The smoke test ran and passed: https://ci.swift.org/view/Pull%20Request/job/swift-PR-osx-smoke-test/1123/

**Comment by eeckstein** [comment]
@najacque ok, thanks for the info. Can you please merge this PR? I cannot do it because it still is in the state that the tests didn't complete.

**Comment by eeckstein** [comment]
Closing as #4277 is already merged
